src/lib/statistics/chunk_statistics/histograms/abstract_histogram.hpp

```diff
@@ -0,0 +1,203 @@
+ * This length is stored in the member `_string_prefix_length`.
+ * Additionally, as of now, we only support a range of ASCII characters, that is stored as a string
+ * in `_supported_characters`. This range must not include gaps and the string has to be sorted.
+ * The possible maximum length of the prefix depends on the number of supported characters.
+ * It must not overflow the uint64_t data type used to represent strings as numbers.
+ * The formula used to verify the prefix length is:
+ * string_prefix_length < std::log(std::numeric_limits<uint64_t>::max()) / std::log(supported_characters.length() + 1)
+ */
+template <typename T>
+class AbstractHistogram : public AbstractFilter {
+ public:
+  AbstractHistogram();
+  AbstractHistogram(const std::string& supported_characters, const uint64_t string_prefix_length);
+  virtual ~AbstractHistogram() = default;
+
+  using HistogramWidthType = std::conditional_t<std::is_same_v<T, std::string>, uint64_t, T>;
+
+  virtual HistogramType histogram_type() const = 0;
+
+  /**
+   * Returns a string with detailed information about the histogram, including the edges of the individual bins.
+   */
+  std::string description() const;
+
+  /**
+   * Returns the estimated selectivity given a predicate type and its parameter(s).
+   * It will always be between 0 and 1.
+   */
+  float estimate_selectivity(const PredicateCondition predicate_type, const AllTypeVariant& variant_value,
+                             const std::optional<AllTypeVariant>& variant_value2 = std::nullopt) const;
+
+  /**
+   * Returns the estimated cardinality given a predicate type and its parameter(s).
+   * It will always be between 0 and total_count().
+   * This method is specialized for strings to handle predicates uniquely applicable to string columns.
+   */
+  float estimate_cardinality(const PredicateCondition predicate_type, const AllTypeVariant& variant_value,
+                             const std::optional<AllTypeVariant>& variant_value2 = std::nullopt) const;
+
+  /**
+   * Returns whether a given predicate type and its parameter(s) can be pruned.
+   * This method is specialized for strings to handle predicates uniquely applicable to string columns.
+   */
+  bool can_prune(const PredicateCondition predicate_type, const AllTypeVariant& variant_value,
+                 const std::optional<AllTypeVariant>& variant_value2 = std::nullopt) const override;
+
+  /**
+   * Returns the lower bound (minimum value) of the histogram.
+   * This is equal to the smallest value in the segment.
+   */
+  T min() const;
+
+  /**
+   * Returns the upper bound (maximum value) of the histogram.
+   * This is equal to the largest value in the segment.
+   */
+  T max() const;
+
+  /**
+   * Given a value return the next representable value.
+   * This method is a wrapper for the functions in histogram_utils.
+   */
+  T get_next_value(const T value) const;
```

Comment:
only used internally, right? Does it need to be public?

Reply:
I used it for the wrapper for `ColumnStatistics`, which is not part of this PR.
Made it protected for now, will look into that again when the time comes.